Sources/NIOHTTPCompression/HTTPDecompression.swift

```diff
@@ -0,0 +1,156 @@
+        }
+
+        var limit: Limit
+
+        /// No limit will be set.
+        public static let none = DecompressionLimit(limit: .none)
+        /// Limit will be set on the request body size.
+        public static func size(_ value: Int) -> DecompressionLimit { return DecompressionLimit(limit: .size(value)) }
+        /// Limit will be set on a ratio between compressed body size and decompressed result.
+        public static func ratio(_ value: Int) -> DecompressionLimit { return DecompressionLimit(limit: .ratio(value)) }
+
+        func exceeded(compressed: Int, decompressed: Int) -> Bool {
+            switch self.limit {
+            case .none:
+                return false
+            case .size(let allowed):
+                return compressed > allowed
+            case .ratio(let ratio):
+                return decompressed > compressed * ratio
+            }
+        }
+    }
+
+    public enum DecompressionError: Error {
+        case limit
+        case inflationError(Int)
+        case initializationError(Int)
+    }
+
+    enum CompressionAlgorithm: String {
+        case gzip
+        case deflate
+
+        init?(header: String?) {
+            switch header {
+            case .some("gzip"):
+                self = .gzip
+            case .some("deflate"):
+                self = .deflate
+            default:
+                return nil
+            }
+        }
+
+        var window: CInt {
+            switch self {
+            case .deflate:
+                return 15
+            case .gzip:
+                return 15 + 16
+            }
+        }
+    }
+}
+
+
+struct Decompressor {
+    enum State {
+        case empty
+        case compressed(NIOHTTPDecompression.CompressionAlgorithm, Int)
+    }
+
+    let limit: NIOHTTPDecompression.DecompressionLimit
+    var state = State.empty
+    var stream = z_stream()
+    var inflated = 0
+
+    public init(limit: NIOHTTPDecompression.DecompressionLimit) {
+        self.limit = limit
+    }
+
+    mutating func decompress(part: inout ByteBuffer, buffer: inout ByteBuffer, originalLength: Int) throws {
+        try self.stream.inflatePart(input: &part, output: &buffer)
+        self.inflated += buffer.readableBytes
+
+        if self.limit.exceeded(compressed: originalLength, decompressed: self.inflated) {
+            throw NIOHTTPDecompression.DecompressionError.limit
+        }
+    }
+
+    mutating func initializeDecoder(encoding: NIOHTTPDecompression.CompressionAlgorithm, length: Int) throws {
+        self.state = .compressed(encoding, length)
+
+        self.stream.zalloc = nil
+        self.stream.zfree = nil
+        self.stream.opaque = nil
+
+        let rc = CNIOExtrasZlib_inflateInit2(&self.stream, encoding.window)
+        guard rc == Z_OK else {
+            throw NIOHTTPDecompression.DecompressionError.initializationError(Int(rc))
+        }
+    }
+
+    mutating func deinitializeDecoder() {
+        inflateEnd(&self.stream)
```

Comment:
Why doesn't this check `.state`? Why doesn't it reset it either?

Reply:
I made `state` be part of the handler, not decompressor, wdyt?

Reply:
That works for me.